examples/Tutorials/modflow6plugin/tutorial02_mf6_plugin.py

```diff
@@ -0,0 +1,147 @@
+# such as nrow*ncol.
+
+
+# define the plugin options
+opt = {
+    "print_input": {"type": "keyword", "optional": "True"},
+    "print_flows": {"type": "keyword", "optional": "True"},
+    "save_input": {"type": "keyword", "optional": "True"},
+    "save_flows": {"type": "keyword", "optional": "True"},
+    "auxiliary": {"type": "string", "shape": "naux", "optional": "True"},
+    "boundnames": {"type": "keyword", "optional": "True"},
+}
+
+# define the plugin stress period data
+# define the stress plugin stress period data
+spv = {
+    "cellid": {"type": "integer", "shape": "ncelldim"},
+    "q_val": {"type": "double precision"},
+    "aux": {"type": "double precision", "shape": "naux", "optional": "True"},
+    "boundname": {"type": "string", "optional": "True"},
+}
+
+# ## Generating FloPy plugin files with FloPy's generate_plugin_template method
+#
+# Once all of your plugin's user defined variables are set up in dictionaries,
+# you can use FloPy's "generate_plugin_template" method to generate the
+# template files for your FloPy plugin.
+#
+# The "generate_plugin_template" method takes the model type (model_type),
+# package abbreviation type (new_package_abbr), options (options), and stress
+# period variables (stress_period_vars) defined above.  Additionally, it can
+# take package variables (package_vars), it can create a template with or
+# without support for the API package (api_package_support), and can also
+# define where you want the main loop of your plugin code to be
+# (evaluation_code_at).
+
+flopy.mf6.utils.flopy_plugins.plugin_template.generate_plugin_template(
+    mt,
+    pn,
+    options=opt,
+    stress_period_vars=spv,
+    api_package_support=True,
+    evaluation_code_at="iteration_start",
+)
+
+# Running generate_plugin_template creates several files in your working
+# directory.  First, a dfn file is created that specifies your plugin's user
+# defined variables.  Then, createpackages.py creates a MODFLOW-6-like
+# package file based on the dfn file, which users of your plugin will
+# instantiate.  A configfpl.py file is also created which tells FloPy how to
+# find your plug-in classes.  Additionally, generate_plugin_template generates
+# a FloPy plugin code file which contains the code that will modify MODFLOW-6's
+# behavior.  Edit this file to change the behavior of your plugin.  Start by
+# adding code in the various callback methods that get called at various times
+# during MODFLOW-6's execution.  These callbacks include init_plugin,
+# receive_bmi, stress_period_start, stress_period_end, time_step_start,
+# time_step_end, iteration_start, iteration_end, and sim_complete.  For
+# example, the code in iteration_start will get called every time MODFLOW-6
+# starts an outer iteration of the solution group the FloPy plugin is
+# operating on.
+#
+# When api_package_support is set to True, the plugin template is set up to
+# use MODFLOW-6's generic API package.  The generic API package can be
```

Comment:
I think we need something like the following.  Not sure where it should go.

MODFLOW 6 has a special type of stress package, called the Application Programming Interface (API) Package.  This package does nothing by itself, but it does provide empty arrays that can be filled from another program that is accessing it through the MODFLOW API.  In the context described here, the plugin transfers stress package information, calculated with Python, into the MODFLOW API Package while MODFLOW is running.  MODFLOW then proceeds with this information and includes it in solution of the matrix equations for a model.